### PR TITLE
Switch RoundUp functions to use a template instead of uint32_t/uint64_t

### PR DESCRIPTION
**Pull request by amzn-tommy**
so that they work with size_t on Mac.

Since size_t is different than both uint32_t and uint64_t on Mac, the original RoundUp functions failed to compile on Mac for places in the code that were calling RoundUp with size_t as input.
-Updated the functions to use a template instead.
-Switched to AZ_MATH_ASSERT, which exists to make it possible to toggle math asserts independently from whether or not it is a profile/debug build
-Removed the divide-by-zero unit tests that were checking for an assert. Since divide-by-zero is undefined behavior, the expected behavior varies based on the build type, and possibly also the platform. As they were before, they would pass in profile on windows, but crash and fail on debug builds.
-Updated the overflow unit tests to only expect an assert on debug builds, and not expect an assert in profile builds, since AZ_MATH_ASSERT is now being used under the hood.

Signed-off-by: Tommy Walton <waltont@amazon.com>

### CODE REVIEW

**Review by rgba16f — APPROVED** [review]

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp [hunk 1]
```diff
@@ -146,32 +151,27 @@ namespace UnitTest
         TestDivideAndRoundUpIsCorrect<uint64_t>();
     }
 
-    
-    class RoundUpInvalidInputTestsFixture : public ScopedAllocatorSetupFixture
-    {
-    };
-
-    TEST_F(RoundUpInvalidInputTestsFixture, DividAndRoundUp_AlignmentZeroUint32_Assert)
```
Comment (Kadino) [inline Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:154]: Recommend adding a comment in this file about intentionally not testing divide by zero due to resulting in undefined behavior. This way someone does not try adding these tests back in.

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 2]
```diff
@@ -467,21 +467,13 @@ namespace AZ
     //! Example: alignment: 4
```
Comment (Kadino) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:467]: Recommend documenting here that 0 is an invalid alignment, which will expose the caller to undefined behavior

**Review by Kadino — APPROVED** [review]
Looks good, some suggestions:

**Review by hultonha — APPROVED** [review]
Looks good and thanks for the thorough description in the PR, always super helpful 👍

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp [hunk 3]
```diff
@@ -180,6 +180,25 @@ namespace UnitTest
         DivideAndRoundUp(
             static_cast<uint64_t>((AZStd::numeric_limits<uint64_t>::max() / 2) + 1),
             static_cast<uint64_t>((AZStd::numeric_limits<uint64_t>::max() / 2) + 1));
+#ifdef AZ_DEBUG_BUILD
+        // AZ_MATH_ASSERT only asserts during debug builds, so expect an assert here
+        AZ_TEST_STOP_TRACE_SUPPRESSION(1);
+#else
+        AZ_TEST_STOP_TRACE_SUPPRESSION(0);
+#endif
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:188]: Instead of having the `#ifdef AZ_DEBUG_BUILD` blocks, `AZ_TEST_STOP_TRACE_SUPPRESSION_NO_COUNT` can be used

Comment (amzn-tommy) [reply Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:188]: The idea is that it should be asserting in debug builds, but not during profile builds. Using NO_COUNT would omit the validation that the assert occurs on invalid input.

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp (hunk 1 above)
Comment (amzn-tommy) [reply Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:154]: Added a comment

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h (hunk 2 above)
Comment (amzn-tommy) [reply Code/Framework/AzCore/AzCore/Math/MathUtils.h:467]: Updated, thanks for the suggestion.
